gobblin-hive-registration/src/main/java/gobblin/hive/HiveSchemaManager.java

```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright (C) 2014-2016 LinkedIn Corp. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+ * this file except in compliance with the License. You may obtain a copy of the
+ * License at  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed
+ * under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+ * CONDITIONS OF ANY KIND, either express or implied.
+ */
+
+package gobblin.hive;
+
+import org.apache.commons.lang3.reflect.ConstructorUtils;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.hive.metastore.api.SerDeInfo;
+
+import com.google.common.base.Enums;
+import com.google.common.base.Optional;
+
+import gobblin.configuration.State;
+import gobblin.hive.avro.HiveAvroSchemaManager;
+import lombok.extern.slf4j.Slf4j;
+
+
+/**
+ * This class manages the schema for Hive registration.
+ *
+ * @author ziliu
+ */
+@Slf4j
+public abstract class HiveSchemaManager {
+
+  protected final State props;
+
+  protected HiveSchemaManager(State props) {
+    this.props = props;
+  }
+
+  /**
+   * Add the appropriate schema properties to the given {@link SerDeInfo}.
+   *
+   * For example, when registering Avro tables, one needs to specify either
+   * {@link org.apache.hadoop.hive.serde2.avro.AvroSerdeUtils.AvroTableProperties#SCHEMA_LITERAL}
+   * or {@link org.apache.hadoop.hive.serde2.avro.AvroSerdeUtils.AvroTableProperties#SCHEMA_URL}
+   * in the {@link SerDeInfo}.
+   */
+  public abstract void addSchemaProperties(SerDeInfo si, Path path);
+
+  public enum Implementation {
+    NOP(HiveNopSchemaManager.class.getName()),
+    AVRO(HiveAvroSchemaManager.class.getName());
+
+    private final String schemaManagerClassName;
+
+    private Implementation(String schemaManagerClassName) {
+      this.schemaManagerClassName = schemaManagerClassName;
+    }
+
+    @Override
+    public String toString() {
+      return this.schemaManagerClassName;
+    }
+  }
+
+  /**
+   * Get an instance of {@link HiveSchemaManager}.
+   *
+   * @param type The {@link HiveSchemaManager} type. It could be AVRO, NOP or the name of a class that implements
+   * {@link HiveSchemaManager}.
+   * @param props A {@link State} object used to instantiate {@link HiveSchemaManager}.
+   * @return
+   */
+  public static HiveSchemaManager getInstance(String type, State props) {
```

Comment:
Consider using a `[ServiceLoader](https://docs.oracle.com/javase/7/docs/api/java/util/ServiceLoader.html)` instead of this pattern. PR https://github.com/linkedin/gobblin/pull/627 has an example of how to use it. This way each implementation of `HiveSchemaManager` does not have to explicitly register itself with this class. It should also allow you to make this class an interface instead of an abstract class.